diff --git a/src/main/java/org/sqlite/core/NativeDB.c b/src/main/java/org/sqlite/core/NativeDB.c
index ed911a9..573f2e1 100644
--- a/src/main/java/org/sqlite/core/NativeDB.c
+++ b/src/main/java/org/sqlite/core/NativeDB.c
@@ -18,7 +18,8 @@
 #include <string.h>
 #include <assert.h>
 #include "NativeDB.h"
-#include "sqlite3.h"
+#include <sqlite3.h>
+#include <spatialite.h>
 
 static jclass dbclass = 0;
 static jclass  fclass = 0;
@@ -212,6 +213,21 @@ static void sethandle(JNIEnv *env, jobject this, sqlite3 * ref)
     (*env)->SetLongField(env, this, pointer, fromref(ref));
 }
 
+static void * getSpatialMemory(JNIEnv *env, jobject this) {
+    static jfieldID pointer = 0;
+    if (!pointer) pointer = (*env)->GetFieldID(env, dbclass, "spatialMemory", "J");
+
+    return toref((*env)->GetLongField(env, this, pointer));
+}
+
+static void setSpatialMemory(JNIEnv *env, jobject this, void* ref) {
+    static jfieldID pointer = 0;
+    if (!pointer) pointer = (*env)->GetFieldID(env, dbclass, "spatialMemory", "J");
+
+    (*env)->SetLongField(env, this, pointer, fromref(ref));
+
+}
+
 struct CollationData {
     JavaVM *vm;
     jobject func;
@@ -542,7 +558,6 @@ JNIEXPORT jint JNICALL Java_org_sqlite_core_NativeDB_enable_1load_1extension(
     return sqlite3_enable_load_extension(db, enable ? 1 : 0);
 }
 
-
 JNIEXPORT void JNICALL Java_org_sqlite_core_NativeDB__1open_1utf8(
         JNIEnv *env, jobject this, jbyteArray file, jint flags)
 {
@@ -574,6 +589,10 @@ JNIEXPORT void JNICALL Java_org_sqlite_core_NativeDB__1open_1utf8(
 
     // Ignore failures, as we can tolerate regular result codes.
     (void) sqlite3_extended_result_codes(db, 1);
+
+    void* spatialMemory = spatialite_alloc_connection();
+    spatialite_init_ex(db, spatialMemory, 1);
+    setSpatialMemory(env, this, spatialMemory);
 }
 
 JNIEXPORT void JNICALL Java_org_sqlite_core_NativeDB__1close(
@@ -588,6 +607,10 @@ JNIEXPORT void JNICALL Java_org_sqlite_core_NativeDB__1close(
         }
         sethandle(env, this, 0);
     }
+    void* spatial = getSpatialMemory(env, this);
+    if (spatial) {
+        spatialite_cleanup_ex(spatial);
+    }
 }
 
 JNIEXPORT void JNICALL Java_org_sqlite_core_NativeDB_interrupt(JNIEnv *env, jobject this)
diff --git a/src/main/java/org/sqlite/core/NativeDB.java b/src/main/java/org/sqlite/core/NativeDB.java
index 3426c2a..010744c 100644
--- a/src/main/java/org/sqlite/core/NativeDB.java
+++ b/src/main/java/org/sqlite/core/NativeDB.java
@@ -25,6 +25,9 @@ import org.sqlite.*;
 public final class NativeDB extends DB {
     /** SQLite connection handle. */
     long pointer = 0;
+    /** Pointer to native memory (see spatialite_init_ex) */
+    long spatialMemory = 0;
+
 
     private static boolean isLoaded;
     private static boolean loadSucceeded;
